File: backend/app/schemas/sequence.py
from datetime import date

from pydantic import BaseModel, Field


class SequenceStateResponse(BaseModel):
    """What the client is allowed to know about today's run.

    `digits` only ever contains the current round's prefix, never the
    full daily sequence.
    """

    challenge_id: int
    challenge_date: date
    round: int
    total_rounds: int
    digits: str | None = None
    completed: bool
    score: int | None = None


class SequenceAnswer(BaseModel):
    round: int = Field(ge=1)
    answer: str = Field(max_length=64)


class SequenceAnswerResponse(BaseModel):
    correct: bool
    round: int
    game_over: bool
    score: int | None = None
    next_round: int | None = None
    next_digits: str | None = None

File: frontend/src/games/sequence/sequence.ts
import { API_BASE_URL } from "../../api/config";

export type SequenceState = {
  challenge_id: number;
  challenge_date: string;
  round: number;
  total_rounds: number;
  // Only ever the digits for the current round, never the full sequence.
  digits: string | null;
  completed: boolean;
  score: number | null;
};

export type SequenceAnswerResponse = {
  correct: boolean;
  round: number;
  game_over: boolean;
  score: number | null;
  next_round: number | null;
  next_digits: string | null;
};

export class SequenceConflictError extends Error {}

export async function getSequenceState(): Promise<SequenceState> {
  const response = await fetch(`${API_BASE_URL}/sequence/state`, {
    credentials: "include",
  });

  if (!response.ok) {
    throw new Error("Couldn't load today's Sequence.");
  }

  return response.json();
}

export async function submitSequenceAnswer(
  round: number,
  answer: string,
): Promise<SequenceAnswerResponse> {
  const response = await fetch(`${API_BASE_URL}/sequence/answer`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    body: JSON.stringify({ round, answer }),
  });

  const data = await response.json();

  if (response.status === 409) {
    // The client and the server disagree about the round. The server wins.
    throw new SequenceConflictError(data.detail);
  }

  if (!response.ok) {
    throw new Error(data.detail ?? "Couldn't submit your answer.");
  }

  return data;
}
